### ISSUE TITLE
docker: Persistent directories are initially empty

### ISSUE THREAD

**Issue by ehlers** [GNS3/gns3-server#512]
When you add a docker container into a project, the persistent directories are empty on first start. This is not a bug of GNS3, it's normal behavior of docker. The host directories are mounted into the container. As the host directory is initially empty, the persistent directory in the container is also empty.

So choosing a directory as persistent has the side effect, that the previous content of that directory in the container is no longer accessible.

That behavior may surprise the users, so it should be clearly documented.

For example the gns3/endhost appliance will not work as expected, as /etc/ssh is used as a persistent directory, which makes the initial content of the /etc/ssh directory inaccessible (moduli, ssh_config, sshd_config). Because of that the ssh daemon will not start.

**Comment by noplay** [comment]
I was thinking about it yesterday and agree that it's make stuff harder.

Perhaps we could do the following:
At boot volumes are mount to /gns3/volumes/volumename (ex:
/gns3/volumes/etc/ssh)
The original dir is renamed
A symlink to the volume is created
If volume is empty, data are copied from orig

We can also totally ignore the volume info in the dockerfile and allow in
settings to persist a specific directory

What do you think?

Le mer. 4 mai 2016 16:07, Bernhard Ehlers <notifications@github.com> a
écrit :

**Comment by ehlers** [comment]
This behavior of persistent volumes is present also in standalone docker, it's not GNS3 specific. So I don't think it needs to be fixed by GNS3. The creator of an docker image can easily handle it himself by saving the contents of the persistent directory into another directory at the end of the Dockerfile. At start of the container the startup-script can copy that back, if the persistent directory is empty.

So I don't want to make GNS3 more complex as it already is, I just want to highlight the need to create a (small) document for developers of docker images about pitfalls they have to take care about.

I like the current use of the volume info as most users won't know, what directories have to be persistent. The creator of the image should have a very good understanding of what's useful. If an advanced user wants to change that behavior he can easily do that by creating his own image based on another base image, that one of the advantages of docker. So I won't change that either.

So in summary I won't change GNS3, just create some small document for image creators.

**Comment by grossmj** [comment]
This will be added to a documentation.
